### DISCUSSION TITLE
Dilating a closed mesh in xy, but not z

### DISCUSSION THREAD

**Discussion by pentacular** [CGAL/cgal#7237] [Q&A]
Hello, I'm trying to dilate a closed mesh in the x and y dimensions, without affecting z.

The implicit distance function approach here looks almost right
https://github.com/CGAL/cgal/blob/master/Polyhedron/demo/Polyhedron/Plugins/Surface_mesh/Offset_meshing_plugin.cpp

But I can't figure out a reasonable way to find the nearest point on the mesh while constraining z.

Any ideas? :)

**Comment by stla** [comment]
You mean an homothety? It's an affine transformation. CGAL allows to apply an affine transformation to a mesh, one just has to enter the matrix.

**Reply by pentacular** [reply]
No, mesh dilation isn't a homothety.

In mesh dilation structures become thicker while remaining stationary.

**Comment by stla** [comment]
That does not *define* a dilation (which is synonym of homothety in geometry).

**Reply by pentacular** [reply]
You may find https://en.wikipedia.org/wiki/Dilation_(morphology) to be a useful reference.

In any case, if you search for "mesh dilation" you will not find people talking about scaling.

**Reply by stla** [reply]
Ah ok. Seen ->[here](https://reference.wolfram.com/language/ref/Dilation.html)<-. That looks related to the Minkowski sum, at least the example with the cube.

**Reply by stla** [reply]
This *is* the Minkowski addition with a circle in the 2D situation ([seen here](https://people.cmm.minesparis.psl.eu/users/serra/cours/pdf/fr/ch2fr.pdf)). Can't you achieve what you want in 3D by Minkowski-adding the mesh of a flat circle to your mesh?

**Reply by stla** [reply]
Here is a xy-non-z-dilated torus.

![dilatedTorus](https://user-images.githubusercontent.com/4466543/217186170-8938f013-7109-448a-a111-37d5b49cbee1.png)

**Reply by stla** [reply]
Clearer with an octahedron:

![dilatedicosahedron](https://user-images.githubusercontent.com/4466543/217187363-9ca6bcc8-3a5d-40ce-968d-e9cde68e5242.png)

**Reply by pentacular** [reply]
Yes, we could convert to nef polyhedron, apply a convex decomposition, apply minkowski summing, and then fuse back together, and convert back again to Surface_mesh.

It's a bit on the convoluted and expensive side for my taste, but it's a possibility.

**Reply by stla** [reply]
Why a convex decomposition? I didn't decompose the torus for example.

Use my R package, straightforward ;-)

**Reply by pentacular** [reply]
Your R package won't be suitable for my use-case.

https://doc.cgal.org/latest/Minkowski_sum_3/group__PkgMinkowskiSum3Ref.html applies a convex decomposition to non-convex inputs, so unless you've re-implemented it, then it's happening behind the scenes.

**Reply by stla** [reply]
Ah yes, behind the scenes. 

Why would there be a use-case for which my package would not be suitable? I'm curious to know more about that. My package is nothing but a wrapper of CGAL.

**Reply by pentacular** [reply]
Well, the uses cases not involving using R come to mind. :)

**Comment by pentacular** [comment]
I wonder if CGAL::Orthogonal_k_neighbor_search would be suitable for this:

https://doc.cgal.org/latest/Spatial_searching/classCGAL_1_1Orthogonal__k__neighbor__search.html

**Comment by sloriot** [comment]
Just thinking out loud: Say you want to modify the offset meshing function with a different distance function. What you could do, is given a query point, take the plane with normal (0,0,1)that contains the query point and compute the distance of that point to the intersection polygons (you can use  [`Polygon_mesh_slicer`](https://doc.cgal.org/latest/Polygon_mesh_processing/classCGAL_1_1Polygon__mesh__slicer.html)). Then this should give something close to what you want (from what I understand).

**Reply by pentacular** [reply]
I think it would work, but wouldn't it be prohibitively expensive to generate slices for every z candidate generated to mesh the implicit function?
